### PR TITLE
Travis-CI build enhancements

### PR DESCRIPTION
**Pull request by mprins**
As discussed in  https://github.com/graphhopper/graphhopper/issues/803#issuecomment-249179426 this PR brings JDK 9 builds on Travis-CI and automatic deployment of snapshots to the central repository.

  - [x] Start testing on Java 9 with Maven 3.3.9
  - [x] upgrade maven plugins and require Maven >= 3.2
  - [x] set up for publishing snapshots to maven central

The latter needs to have the secure keys generated and added in the `.travis.yml` (there are some placeholders there now) so that after a succesful build of the master branch the snapshot artifacts are deployed.

This build no longer uses the `/core/files/travis-build.sh` not sure if that can be removed, it has some out-commented npm stuff.

### CODE REVIEW

**Comment by karussell** [comment]
Already really nice - thanks!

**Comment by mprins** [comment]
@karussell this needs a little something on your side to add the secure keys to the travis file; deployment should the start working after merging into master.

Also I've set up to allow test failures on java 9, and it's showing some with a message of `unable to unmap the mapped buffer`. eg. https://travis-ci.org/graphhopper/graphhopper/jobs/162408743#L1011

**Comment by karussell** [comment]
Thanks - will read next week about how to setup the secure keys.

Regarding the JDK9: in #701 I already stumbled over this ugliness, but I'm unsure how to fix this yet (see the link there). Maybe we can exclude or ignore (for now) a single test for a specific JDK version?

File: .travis.yml [hunk 1]
```diff
@@ -1,12 +1,70 @@
...
+     fi
+   - if [ "$TRAVIS_JDK_VERSION" == oraclejdk9 ]; then
+         sudo rm /etc/mavenrc;
+     fi
+
 install: true
-script: ./core/files/travis-build.sh
+
+script:
+  - mvn clean test verify -B
```
Comment (karussell) [inline .travis.yml]: I've looked up what the -B means (--batch-mode) ... why is this necessary?

File: core/pom.xml [hunk 2]
```diff
@@ -74,41 +74,43 @@
...
-                </plugin>                
-                
-                <!-- create jar with test classes to be reused in other projects like external or our reader-osm -->                
+                </plugin>
+
+                <!-- create jar with test classes to be reused in other projects like external or our reader-osm -->
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-jar-plugin</artifactId>
+                    <version>3.0.2</version>
```
Comment (karussell) [inline core/pom.xml]: As you already changed this in the parent pom, this is not necessary here (?) and when not being explicit here, we can then easier upgrade versions later on in the parent pom

File: web/pom.xml [hunk 3]
```diff
@@ -80,22 +80,23 @@
...
+
     </dependencies>
 
     <build>
         <plugins>
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-compiler-plugin</artifactId>                
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.5.1</version>
```
Comment (karussell) [inline web/pom.xml]: Same here: we could remove the version (?)

File: .travis.yml (hunk 1 above)
Comment (mprins) [inline .travis.yml]: it would prevent the build waiting until it times out if any interactive response would be required

File: core/pom.xml (hunk 2 above)
Comment (mprins) [inline core/pom.xml]: that would only work if the parent pom has a pluginManagement section where the version is specified otherwise the result of the version resolution of the plugin is undefined (maven will just pick one that's available), see: https://maven.apache.org/pom.html#Plugin_Management

File: web/pom.xml (hunk 3 above)
Comment (mprins) [inline web/pom.xml]: as above, not really a good idea

File: core/pom.xml (hunk 2 above)
Comment (karussell) [inline core/pom.xml]: Ups, thanks for pointing this out :) ! Maybe it would then be better to have a plugin section in the root pom or are there disadvantages?

**Comment by karussell** [comment]
Uhm, strange. How it comes I have a comment in my emails regarding the 2 profiles but not here?

> You could specify two profiles, one for jdk8 and one for jdk9...

**Comment by karussell** [comment]
Nice! Deployment seems to work (according to the time stamps of the snapshot) but still something timed out: https://travis-ci.org/graphhopper/graphhopper/jobs/163212260

>   travis_jigger $! $timeout $cmd

Before I can merge it: would you mind to sign the [CLA](https://github.com/graphhopper/graphhopper/blob/master/.github/CONTRIBUTING.md) (you do not give away your rights, just agree to the Apache License) and send this to us. Or just send me your email so that I can send you an easier to handle digital signing doc.

**Comment by devemux86** [comment]
>travis_jigger $! $timeout $cmd

That probably comes from [travis_wait](https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received), is it needed (default 10min)?

File: core/pom.xml (hunk 2 above)
Comment (mprins) [inline core/pom.xml]: You can setup the plugins in the parent pom like:
```xml
<build>
    <pluginManagement>
            <plugins>
                <plugin>
                    <groupId>org.apache.maven.plugins</groupId>
                    <artifactId>maven-compiler-plugin</artifactId>
                    <version>3.5.1</version>
                    <configuration>
                        <source>${java.version}</source>
                        <target>${java.version}</target>
                    </configuration>
                </plugin>
...
            </plugins>
        </pluginManagement>
```
and then in the build section of the child projects and possibly the parent as well specify the plugins to be used for the build of that artifact. You can find a working example in https://github.com/B3Partners/flamingo-ibis/blob/master/pom.xml

the downside is that there is some duplication of plugins (they both the `pluginManagement` and the `build` section of a pom will have a `plugins` section)

**Comment by mprins** [comment]
the `travis_wait` may not be necessary, if there is output within the 10 mins. if the deploy produces no log output for 10min or more travis will fail the deploy as it seems stalled (to travis)

**Comment by mprins** [comment]
>Uhm, strange. How it comes I have a comment in my emails regarding the 2 profiles but not here?
>>You could specify two profiles, one for jdk8 and one for jdk9...

@karussell I commented on the original issue #701

**Comment by karussell** [comment]
Merged - thanks a lot again @mprins - really nice now :) !

We'll improve this now in an iterative manner later on ...

**Comment by karussell** [comment]
Currently the snapshots are not deployed, e.g. see https://travis-ci.org/graphhopper/graphhopper/jobs/168299877

It says: 
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:2.7:deploy (default-deploy) on project graphhopper-parent: Failed to deploy artifacts: Could not transfer artifact com.graphhopper:graphhopper-parent:pom:0.9-20161017.130625-2 from/to ossrh (https://oss.sonatype.org/content/repositories/snapshots): Failed to transfer file: https://oss.sonatype.org/content/repositories/snapshots/com/graphhopper/graphhopper-parent/0.9-SNAPSHOT/graphhopper-parent-0.9-20161017.130625-2.pom. Return code is: 401, ReasonPhrase: Unauthorized. -> [Help 1]
```

Any clue what changed?

Also it says:
```
This job is running on container-based infrastructure, which does not allow use of 'sudo', setuid and setguid executables.

If you require sudo, add 'sudo: required' to your .travis.yml
```

And we have `sudo: false` in the root but in the matrix we have `sudo: required`. Is this correct?

**Comment by mprins** [comment]
> Any clue about this?
The username and password are invalid; looking at the .travis.yml the are:
```
    - secure: "CI_DEPLOY_USERNAME=central_username"
    - secure: "CI_DEPLOY_PASSWORD=central_password"
```

those still need to be update to your (encrypted) credentials as I wrote up top

>> The latter needs to have the secure keys generated and added in the `.travis.yml` (there are some placeholders there now) so that after a successful build of the master branch the snapshot artifacts are deployed.

see: https://docs.travis-ci.com/user/environment-variables/#Defining-encrypted-variables-in-.travis.yml (and https://docs.travis-ci.com/user/encryption-keys )

ps When using the Travis gem I would not use the --add option, it will ruin the `.travis.yml` layout


> And we have sudo: false in the root but in the matrix section we have sudo: required. Is this correct?

We are changing to a different build image as part of the build matrix; using 
```yaml
sudo: required
dist: trusty
group: edge
```
the sudo is required to update the java 9 jdk on that machine to a current version (there is no relation to the snapshot deployment failure)

**Comment by karussell** [comment]
> those still need to be update to your (encrypted) credentials as I wrote up top

I think this is a merge issue (or something else) as I've definitely updated them - thanks for pointing me to this :) !

> ps When using the Travis gem I would not use the --add option, it will ruin the .travis.yml layout

what do you mean here?

> We are changing to a different build image as part of the build matrix; using 

Thanks for the explanation!

**Comment by karussell** [comment]
Works now - thanks again @mprins :) ! E.g. see [here](https://oss.sonatype.org/content/groups/public/com/graphhopper/graphhopper-core/0.9-SNAPSHOT/)

**Comment by karussell** [comment]
Just as a reference there is a web UI to create those keys: http://rkh.github.io/travis-encrypt/public/index.html
